src/main/java/org/apache/flink/benchmark/AsyncWaitOperatorBenchmark.java

```diff
@@ -0,0 +1,120 @@
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.datastream.AsyncDataStream;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.DataStreamSource;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.async.ResultFuture;
+import org.apache.flink.streaming.api.functions.async.RichAsyncFunction;
+import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+
+import java.util.Collections;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+@OperationsPerInvocation(value = AsyncWaitOperatorBenchmark.RECORDS_PER_INVOCATION)
+public class AsyncWaitOperatorBenchmark extends BenchmarkBase {
+	public static final int RECORDS_PER_INVOCATION = 1_000_000;
+
+	private static final long CHECKPOINT_INTERVAL_MS = 100;
+
+	@Param
+	public AsyncDataStream.OutputMode outputMode;
+
+	public static void main(String[] args)
+		throws RunnerException {
+		Options options = new OptionsBuilder()
+			.verbosity(VerboseMode.NORMAL)
+			.include(".*" + AsyncWaitOperatorBenchmark.class.getSimpleName() + ".*")
+			.build();
+
+		new Runner(options).run();
+	}
+
+	@Benchmark
+	public void asyncWait(FlinkEnvironmentContext context) throws Exception {
+
+		StreamExecutionEnvironment env = context.env;
+		env.enableCheckpointing(CHECKPOINT_INTERVAL_MS);
+		env.setParallelism(1);
+
+		DataStreamSource<Long> source = env.addSource(new LongSource(RECORDS_PER_INVOCATION));
+		DataStream<Long> result = createAsyncOperator(source);
+		result.addSink(new DiscardingSink<>());
+
+		env.execute();
+	}
+
+	private DataStream<Long> createAsyncOperator(DataStreamSource<Long> source) {
+		switch (outputMode) {
+			case ORDERED:
+				return AsyncDataStream.orderedWait(
+						source,
+						new BenchmarkAsyncFunctionExecutor(),
+						0,
+						TimeUnit.MILLISECONDS);
+			case UNORDERED:
+				return AsyncDataStream.unorderedWait(
+						source,
+						new BenchmarkAsyncFunctionExecutor(),
+						0,
+						TimeUnit.MILLISECONDS);
+			default:
+				throw new UnsupportedOperationException("Unknown mode");
+		}
+	}
+
+	private static class BenchmarkAsyncFunctionExecutor extends RichAsyncFunction<Long, Long> {
+		private transient ExecutorService executor;
+
+		@Override
+		public void open(Configuration parameters) throws Exception {
+			super.open(parameters);
+
+			executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
```

Comment:
Instead of starting up the thread pool and shutting it down once per iteration in the `open()`, we could:
- move this to `setUp()` and `teraDown()` methods per iteration, but wouldn't be measured
- move this to `setUp()` and `teraDown()` methods per JVM instance (I'm not sure but I would guess JMH supports that. If not, we could make it just a static variable).

That would reduce the startup overhead by a bit, however I don't know if in any measurable way?

Reply:
👍